File: RaylibPlatformer/RaylibPlatformer/Player.cpp
#include"Player.h"

Player::Player() {}
Player::Player(float _width, float _height, Texture2D _tex, Vector2 _pos, PhysicsAttributes _phys, float _moveSpeed, float _jumpSpeed)
{
	#pragma region Physics

	width = _width;
	height = _height;
	pos = _pos;
	phys = _phys;
	moveSpeed = _moveSpeed;
	jumpSpeed = _jumpSpeed;

	#pragma endregion

	#pragma region Animation

	//Initializing animations with slicing and playSpeed data
	idleAnimation = Animation(_tex, 1, { 1,1 }, { 1,1 }, { 18,18 }, 0, 0);
	runAnimation = Animation(_tex, 8, { 1,1 }, { 1,1 }, { 18,18 }, 15, 1);
	jumpAnimation = Animation(_tex, 1, { 1,1 }, { 1,1 }, { 18,18 }, 0, 2);
	fallAnimation = Animation(_tex, 1, { 1,1 }, { 1,1 }, { 18,18 }, 0, 3);
	wallJumpAnimation = Animation(_tex, 1, { 1,1 }, { 1,1 }, { 18,18 }, 0, 4);

	//Initializing animation transitions (transitions occur from any state)
	animationController.transitions.push_back({ false, idleAnimation });
	animationController.transitions.push_back({ false, runAnimation });
	animationController.transitions.push_back({ false, jumpAnimation });
	animationController.transitions.push_back({ false, fallAnimation });
	animationController.transitions.push_back({ false, wallJumpAnimation });

	//Assigining the default animation
	animationController.currentAnimation = &idleAnimation;

	//Texture2D smokePuffTex = LoadTexture("Assets/Textures/smokeParticle.png");
	//smokePuff = new Particle(smokePuffTex, { 100,100 }, WHITE, Animation(smokePuffTex, 6, { 1,1 }, { 1,1 }, { 8,8 }, 30, 0), true);

	#pragma endregion
}
Player::~Player()
{
	//delete smokePuff;
}

void Player::Update(Rectangle levelColliders[], int arraySize)
{
#pragma region Input

	if ((joystick ? IsGamepadButtonDown(GAMEPAD_PLAYER1, GAMEPAD_XBOX_BUTTON_LEFT) : IsKeyDown(KEY_A)) && canMoveLeft) phys.velocity.x -= moveSpeed;
	if ((joystick ? IsGamepadButtonDown(GAMEPAD_PLAYER1, GAMEPAD_XBOX_BUTTON_RIGHT) : IsKeyDown(KEY_D)) && canMoveRight) phys.velocity.x += moveSpeed;

	if (joystick ? IsGamepadButtonPressed(GAMEPAD_PLAYER1, GAMEPAD_XBOX_BUTTON_A) : IsKeyPressed(KEY_SPACE))
	{
		if (grounded)
		{
			pos.y--; //Separates collider from ground
			phys.velocity.y = -jumpSpeed;
		}
		else if (!canMoveLeft)
		{
			phys.velocity.x += moveSpeed * 10;
			pos.y--; //Separates collider from ground
			phys.velocity.y = -jumpSpeed;
		}
		else if (!canMoveRight)
		{
			phys.velocity.x -= moveSpeed * 10;
			pos.y--; //Separates collider from ground
			phys.velocity.y = -jumpSpeed;
		}
	}
	if ((joystick ? IsGamepadButtonReleased(GAMEPAD_PLAYER1, GAMEPAD_XBOX_BUTTON_A) : IsKeyReleased(KEY_SPACE)) && phys.velocity.y < 0)
	{
		phys.velocity.y /= 2;
	}

#pragma endregion

#pragma region Physics/Collision

	phys.Update();

	//Prevents Y velocity from exceeding maximum
	if (phys.velocity.y > phys.maxVelocity.y) phys.velocity.y = phys.maxVelocity.y;

	pos.x += phys.velocity.x;
	pos.y += phys.velocity.y;

	UpdateColliders();
	CheckColliders(levelColliders, arraySize);

#pragma endregion

#pragma region Animation

	animationController.transitions[0].condition = (abs(phys.velocity.x) <= 0.3f && grounded); //Idle
	animationController.transitions[1].condition = (abs(phys.velocity.x) > 0.3f && grounded); //Run
	animationController.transitions[2].condition = (phys.velocity.y < 0 && !grounded); //Jump
	animationController.transitions[3].condition = (phys.velocity.y > 0 && !grounded); //Fall
	if (!grounded && !canMoveLeft) //Left walljump
	{
		flipX = false;
		animationController.transitions[4].condition = true;
		phys.gravity = 0.15f;
	}
	else if (!grounded && !canMoveRight) //Right walljump
	{
		flipX = true;
		animationController.transitions[4].condition = true;
		phys.gravity = 0.15f;
	}
	else
	{
		animationController.transitions[4].condition = false;
		if (phys.velocity.x < 0) flipX = true;
		if (phys.velocity.x > 0) flipX = false;
		phys.gravity = 0.2f;
	}

	//smokePuff->Update();

#pragma endregion

}

void Player::Draw(Vector2 cameraOffset, Vector2 screenOffset)
{
	//DrawCollisionRecs();
	animationController.Update();
	animationController.DrawAnimation({ pos.x - 1, pos.y - 2 }, WHITE, flipX, flipY, rotation, cameraOffset, screenOffset);
	//smokePuff->Draw(cameraOffset, screenOffset);
}

void Player::DrawCollisionRecs()
{
	DrawRectangleRec(collisionRecs[0], DARKGRAY);
	DrawRectangleRec(collisionRecs[1], RED);
	DrawRectangleRec(collisionRecs[2], BLUE);
	DrawRectangleRec(collisionRecs[5], PINK);
	DrawRectangleRec(collisionRecs[6], ORANGE);
	DrawRectangleRec(collisionRecs[3], GREEN);
	DrawRectangleRec(collisionRecs[4], YELLOW);
}

void Player::UpdateColliders()
{
	collisionRecs[0] = { pos.x, pos.y, width, height }; //Center
	collisionRecs[1] = { pos.x + 4, pos.y - 1, width - 8, 1 }; //Top
	collisionRecs[2] = { pos.x + 4, pos.y + height, width - 8, 1 }; //Bottom
	collisionRecs[3] = { pos.x, pos.y + 4, 1, height - 8 }; //Left
	collisionRecs[4] = { pos.x + width - 2, pos.y + 4, 1, height - 8 }; //Right
	collisionRecs[5] = { pos.x - 1, pos.y + 4, 2, height - 8}; //Left input catch
	collisionRecs[6] = { pos.x + width - 2, pos.y + 4, 2, height - 8}; //Right input catch
}

void Player::CheckColliders(Rectangle levelColliders[], int arraySize)
{
	grounded = false;
	canMoveLeft = true;
	canMoveRight = true;

	for (int i = 0; i < arraySize; i++)
	{
		if (CheckCollisionRecs(collisionRecs[1], levelColliders[i])) //Top
		{
			pos.y += (levelColliders[i].y + levelColliders[i].height + 1) - pos.y;
			phys.velocity.y = 0;
		}
		if (CheckCollisionRecs(collisionRecs[2], levelColliders[i])) //Bottom
		{
			pos.y -= ((pos.y + height) - levelColliders[i].y);
			phys.velocity.y = 0;
		}
		if (CheckCollisionRecs(collisionRecs[3], levelColliders[i])) //Left
		{
			pos.x++;
			phys.velocity.x = 0;
		}
		if (CheckCollisionRecs(collisionRecs[4], levelColliders[i])) //Right
		{
			pos.x--;
			phys.velocity.x = 0;
		}
		//Additional check for bottom collider + one additional pixel below for jumping
		if (CheckCollisionRecs({ pos.x + 4, pos.y + height - 2, width - 8, 2 }, levelColliders[i])) grounded = true;
		//Left collider to prevent input
		if (CheckCollisionRecs(collisionRecs[5], levelColliders[i])) canMoveLeft = false;
		//Right collider to prevent input
		if (CheckCollisionRecs(collisionRecs[6], levelColliders[i])) canMoveRight = false;
	}
}

File: CSharpPlatformer/Platformer/Player.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Raylib;
using static Raylib.Raylib;

namespace Platformer
{
    public class Player
    {
        //General
        public Vector2 pos;
        public Vector2 size;

        //Physics
        public PhysicsAttributes physAtts;
        public float moveSpeed;
        public float jumpSpeed;
        public Rectangle[] collisionRecs = new Rectangle[8];
        public bool grounded;
        public bool initialJump;
        public bool canWallJumpLeft;
        public bool canWallJumpRight;
        public float wallJumpForce;

        //Graphics
        public Texture2D tex;
        public Color tint;
        bool flipX;
        bool flipY;

        AnimationController animationController;
        Animation idleAnimation;
        Animation runAnimation;
        Animation jumpAnimation;
        Animation fallAnimation;
        Animation wallJumpAnimation;

        public Player(Vector2 startPos)
        {
            //General Initialization
            pos = startPos;
            size = new Vector2(16, 16);

            //Physics Attributes
            physAtts = new PhysicsAttributes();
            physAtts.drag = 0.75f;
            physAtts.gravity = 0.2f;
            physAtts.velocity = Vector2.Zero;
            physAtts.maxVelocity = new Vector2(2, 5);

            //Other Physics
            moveSpeed = 0.75f;
            jumpSpeed = 5f;
            grounded = false;
            initialJump = false;
            canWallJumpLeft = false;
            canWallJumpRight = false;
            wallJumpForce = 2;

            //Graphics/Animation
            tex = LoadTexture("assets/playerSheet.png");
            tint = Color.WHITE;
            flipX = false;
            flipY = false;

            Animation idleAnimation     = new Animation(tex, 1, new Vector2(1, 1), new Vector2(1, 1), new Vector2(18, 18), 0,  0);
            Animation runAnimation      = new Animation(tex, 8, new Vector2(1, 1), new Vector2(1, 1), new Vector2(18, 18), 15, 1);
            Animation jumpAnimation     = new Animation(tex, 1, new Vector2(1, 1), new Vector2(1, 1), new Vector2(18, 18), 0,  2);
            Animation fallAnimation     = new Animation(tex, 1, new Vector2(1, 1), new Vector2(1, 1), new Vector2(18, 18), 0,  3);
            Animation wallJumpAnimation = new Animation(tex, 1, new Vector2(1, 1), new Vector2(1, 1), new Vector2(18, 18), 0,  4);

            animationController = new AnimationController();
            animationController.transitions.Add(new AnimationTransition(false, idleAnimation));
            animationController.transitions.Add(new AnimationTransition(false, runAnimation));
            animationController.transitions.Add(new AnimationTransition(false, jumpAnimation));
            animationController.transitions.Add(new AnimationTransition(false, fallAnimation));
            animationController.transitions.Add(new AnimationTransition(false, wallJumpAnimation));
            animationController.currentAnimation = idleAnimation;

            UpdateColliders();
        }

        public void Update(Rectangle[] levelColliders)
        {
            #region Input

            if (IsKeyDown(KeyboardKey.KEY_A) || 
                IsGamepadButtonDown(GamepadNumber.GAMEPAD_PLAYER1, GamepadButton.GAMEPAD_BUTTON_LEFT_FACE_LEFT))
                physAtts.velocity.x -= moveSpeed;

            if (IsKeyDown(KeyboardKey.KEY_D) || 
                IsGamepadButtonDown(GamepadNumber.GAMEPAD_PLAYER1, GamepadButton.GAMEPAD_BUTTON_LEFT_FACE_RIGHT))
                physAtts.velocity.x += moveSpeed;

            if (IsKeyPressed(KeyboardKey.KEY_SPACE) ||
                IsGamepadButtonPressed(GamepadNumber.GAMEPAD_PLAYER1, GamepadButton.GAMEPAD_BUTTON_RIGHT_FACE_DOWN))
            {
                if (grounded)
                {
                    physAtts.velocity.y = -jumpSpeed;
                    initialJump = true;
                }
                else if (canWallJumpLeft)
                {
                    physAtts.velocity.x = wallJumpForce;
                    physAtts.velocity.y = -jumpSpeed;
                    initialJump = true;
                }
                else if (canWallJumpRight)
                {
                    physAtts.velocity.x = -wallJumpForce;
                    physAtts.velocity.y = -jumpSpeed;
                    initialJump = true;
                }
            }

            if (initialJump && physAtts.velocity.y < 0 && (IsKeyReleased(KeyboardKey.KEY_SPACE) ||
                IsGamepadButtonReleased(GamepadNumber.GAMEPAD_PLAYER1, GamepadButton.GAMEPAD_BUTTON_RIGHT_FACE_DOWN)))
            {
                physAtts.velocity.y /= 2;
                initialJump = false; //TODO ensure the player cannot start fast falling when they're not supposed to
            }

            #endregion

            #region Physics

            physAtts.Update(grounded);
            //Prevents velocity from exceeding the maximum
            if (physAtts.velocity.y > physAtts.maxVelocity.y) physAtts.velocity.y = physAtts.maxVelocity.y;
            if (physAtts.velocity.x < -physAtts.maxVelocity.x) physAtts.velocity.x = -physAtts.maxVelocity.x;
            if (physAtts.velocity.x > physAtts.maxVelocity.x) physAtts.velocity.x = physAtts.maxVelocity.x;
            //Adds the velocity to the player's position
            pos += physAtts.velocity;
            //Adjust colliders to their new positions
            UpdateColliders();
            //Check for collision and correct physics accordingly
            CheckPhysicsColliders(levelColliders);

            #endregion

            #region Animation

            animationController.transitions[0].condition = (Math.Abs(physAtts.velocity.x) <= 0.3f && grounded); //Idle
            animationController.transitions[1].condition = (Math.Abs(physAtts.velocity.x) > 0.3f && grounded); //Run
            animationController.transitions[2].condition = (physAtts.velocity.y < 0 && !grounded); //Jump
            animationController.transitions[3].condition = (physAtts.velocity.y > 0 && !grounded); //Fall

            if (canWallJumpLeft) //Left Wall Jump
            {
                flipX = false;
                animationController.transitions[4].condition = true;
            }
            else if (canWallJumpRight) //Right Wall Jump
            {
                flipX = true;
                animationController.transitions[4].condition = true;
            }
            else
            {
                animationController.transitions[4].condition = false;
                if (physAtts.velocity.x < 0) flipX = true;
                if (physAtts.velocity.x > 0) flipX = false;
            }

            #endregion
        }

        public void UpdateColliders()
        {
            collisionRecs[0] = new Rectangle(pos.x, pos.y, size.x, size.y);             // Center
            collisionRecs[1] = new Rectangle(pos.x + 4, pos.y - 1, size.x - 8, 1);      // Top
            collisionRecs[2] = new Rectangle(pos.x + 4, pos.y + size.y, size.x - 8, 1); // Bottom
            collisionRecs[3] = new Rectangle(pos.x + 1, pos.y + 4, 1, size.y - 8);      // Left
            collisionRecs[4] = new Rectangle(pos.x + size.x - 2, pos.y + 4, 1, size.y - 8); // Right
            collisionRecs[5] = new Rectangle(pos.x + 4, pos.y + size.y, size.x - 8, 2); // Ground Check (floor)
            collisionRecs[6] = new Rectangle(pos.x - 1, pos.y + 4, 2, size.y - 8);      // Left Wall Jump
            collisionRecs[7] = new Rectangle(pos.x + size.x - 1, pos.y + 4, 2, size.y - 8); // Right Wall Jump
        }

        public void CheckPhysicsColliders(Rectangle[] otherRecs)
        {
            grounded = false;
            canWallJumpLeft = false;
            canWallJumpRight = false;

            for (int i = 0; i < otherRecs.Length; i++)
            {
                if (physAtts.velocity.y < 0 && CheckCollisionRecs(collisionRecs[1], otherRecs[i])) //Top
                {
                    pos.y += (otherRecs[i].y + otherRecs[i].height + 1) - pos.y;
                    physAtts.velocity.y = 0;
                }
                if (physAtts.velocity.y > 0 && CheckCollisionRecs(collisionRecs[2], otherRecs[i])) //Bottom
                {
                    pos.y -= ((pos.y + size.y) - otherRecs[i].y);
                    physAtts.velocity.y = 0;
                }
                if (CheckCollisionRecs(collisionRecs[3], otherRecs[i])) //Left
                {
                    pos.x++;
                    physAtts.velocity.x = 0;
                }
                if (CheckCollisionRecs(collisionRecs[4], otherRecs[i])) //Right
                {
                    pos.x--;
                    physAtts.velocity.x = 0;
                }
                if (CheckCollisionRecs(collisionRecs[5], otherRecs[i])) //Ground Check
                    grounded = true;
                if (!grounded && CheckCollisionRecs(collisionRecs[6], otherRecs[i])) //Left Wall Jump Check
                    canWallJumpLeft = true;
                if (!grounded && CheckCollisionRecs(collisionRecs[7], otherRecs[i])) //Right Wall Jump Check
                    canWallJumpRight = true;
            }
        }

        public void Draw(Vector2 cameraOffset, Vector2 screenOffset)
        {
            animationController.Update();
            animationController.Draw(new Vector2(pos.x - 1, pos.y - 2), tint, flipX, flipY, size / 2, 0, cameraOffset, screenOffset);
        }
        public void DrawColliders()
        {
            Color[] debugColors = new Color [] { Color.GRAY, Color.RED, Color.BLUE, Color.YELLOW, Color.GREEN, Color.BLANK, Color.BLANK, Color.BLANK };
            for (int i = 0; i < collisionRecs.Length; i++)
                DrawRectangle((int)collisionRecs[i].x, (int)collisionRecs[i].y, (int)collisionRecs[i].width, (int)collisionRecs[i].height, debugColors[i]);
        }
    }
}
